src/transformers/models/wav2vec2/modeling_wav2vec2.py

```diff
@@ -669,6 +667,256 @@ def custom_forward(*inputs):
+
+        vars = self.vars
+        if self.combine_groups:
+            vars = vars.repeat(1, self.groups, 1)
+
+        if produce_targets:
+            targets = x.view(bsz * tsz * self.groups, -1).argmax(dim=-1).view(bsz, tsz, self.groups).detach()
+            print(f"VQ targets shape: {targets.shape}")
+
+        x = x.unsqueeze(-1) * vars
+        x = x.view(bsz * tsz, self.groups, self.num_vars, -1)
+        x = x.sum(-2)
+        x = x.view(bsz, tsz, -1)
+
+        if not self.time_first:
+            x = x.transpose(1, 2)  # BTC -> BCT
+
+        return x, num_vars, prob_perplexity
+
+
+class Wav2Vec2Quantizer(nn.Module):
+    def __init__(self, config: Wav2Vec2Config):
+        super().__init__()
+        self.config = config
+
+        final_dim = config.final_dim if config.final_dim > 0 else config.conv_dim[-1]
+        vq_dim = config.latent_dim if config.latent_dim > 0 else final_dim
+        self.quantizer = GumbelVectorQuantizer(
+            dim=config.conv_dim[-1],
+            num_vars=config.num_latent_vars,
+            temp=config.latent_temp,
+            groups=config.num_latent_groups,
+            combine_groups=False,
+            vq_dim=vq_dim,
+            time_first=True,
+        )
+        self.project_q = nn.Linear(vq_dim, final_dim)
+        self.final_proj = nn.Linear(config.hidden_size, final_dim)
+
+        self.num_negatives = config.num_negatives
+        self.cross_sample_negatives = config.cross_sample_negatives
+
+    def forward(self, extractor_hidden_states, transformer_hidden_states, mask_time_indices=None):
+        if mask_time_indices is not None:
+            # quantize only hidden states in places of masks
+            extractor_hidden_states = extractor_hidden_states[mask_time_indices].view(
+                extractor_hidden_states.size(0), -1, extractor_hidden_states.size(-1)
+            )
+
+        quantized_states, num_vars, prob_perplexity = self.quantizer(extractor_hidden_states, produce_targets=False)
+        quantized_states = self.project_q(quantized_states)
+
+        negatives, _ = self.sample_negatives(quantized_states, quantized_states.size(1))
+        negatives = self.project_q(negatives)
+        transformer_hidden_states = self.final_proj(transformer_hidden_states)
+
+        return transformer_hidden_states, quantized_states, negatives, num_vars, prob_perplexity
+
+    def sample_negatives(self, y, num, padding_count=None):
+
+        if self.num_negatives == 0 and self.cross_sample_negatives == 0:
+            return y.new(0)
+
+        bsz, tsz, fsz = y.shape
+        y = y.view(-1, fsz)  # BTC => (BxT)C
+
+        # FIXME: what happens if padding_count is specified?
+        cross_high = tsz * bsz
+        high = tsz - (padding_count or 0)
+        with torch.no_grad():
+            assert high > 1, f"{bsz, tsz, fsz}"
+
+            if self.num_negatives > 0:
+                tszs = torch.arange(num).unsqueeze(-1).expand(-1, self.num_negatives).flatten()
```

Comment:
this creates a new tensor - do we need to be careful about the device here? E.g. should we add a `.to(y.device)`?

Reply:
The created tensors that are left in the current version are only used for indexing and then discarded, so the device shouldn't matter much (I've checked both cpu and cuda versions)